osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs

```diff
@@ -264,70 +265,93 @@ private PathType convertPathType(string input)
+            string[] pointStringSplit = pointString.Split('|');
 
-            var controlPoints = new List<Memory<PathControlPoint>>();
-            int startIndex = 0;
-            int endIndex = 0;
-            bool first = true;
+            var pointsBuffer = ArrayPool<Vector2>.Shared.Rent(pointStringSplit.Length);
+            var segmentsBuffer = ArrayPool<(PathType Type, int StartIndex)>.Shared.Rent(pointStringSplit.Length);
+            int currentPointsIndex = 0;
+            int currentSegmentsIndex = 0;
 
-            while (++endIndex < pointSplit.Length)
+            try
             {
-                // Keep incrementing endIndex while it's not the start of a new segment (indicated by having an alpha character at position 0).
-                if (!char.IsLetter(pointSplit[endIndex][0]))
-                    continue;
+                foreach (string s in pointStringSplit)
+                {
+                    if (char.IsLetter(s[0]))
+                    {
+                        // The start of a new segment(indicated by having an alpha character at position 0).
+                        var pathType = convertPathType(s);
+                        segmentsBuffer[currentSegmentsIndex++] = (pathType, currentPointsIndex);
+
+                        // First segment is prepended by an extra zero point
+                        if (currentPointsIndex == 0)
+                            pointsBuffer[currentPointsIndex++] = Vector2.Zero;
+                    }
+                    else
+                    {
+                        pointsBuffer[currentPointsIndex++] = readPoint(s, offset);
+                    }
+                }
 
-                // Multi-segmented sliders DON'T contain the end point as part of the current segment as it's assumed to be the start of the next segment.
-                // The start of the next segment is the index after the type descriptor.
-                string endPoint = endIndex < pointSplit.Length - 1 ? pointSplit[endIndex + 1] : null;
+                int pointsCount = currentPointsIndex;
+                int segmentsCount = currentSegmentsIndex;
+                var controlPoints = new List<ArraySegment<PathControlPoint>>(pointsCount);
+                var allPoints = new ArraySegment<Vector2>(pointsBuffer, 0, pointsCount);
 
-                controlPoints.AddRange(convertPoints(pointSplit.AsMemory().Slice(startIndex, endIndex - startIndex), endPoint, first, offset));
-                startIndex = endIndex;
-                first = false;
+                for (int i = 0; i < segmentsCount; i++)
+                {
+                    if (i < segmentsCount - 1)
+                    {
+                        int startIndex = segmentsBuffer[i].StartIndex;
+                        int endIndex = segmentsBuffer[i + 1].StartIndex;
+                        controlPoints.AddRange(convertPoints(segmentsBuffer[i].Type, allPoints.Slice(startIndex, endIndex - startIndex), pointsBuffer[endIndex]));
```

Comment:
Specifically, I think *this* particular case is better readable with range expression.

I don't have preference on other cases though.

Reply:
There have been multiple occurrences of the `..` operator being used and us not liking it in review, I don't want that happening here so I changed it with the `Slice` method to move the PR forward.